### ISSUE TITLE
GoCD agent config files require CRLF line endings (under Windows?)

### ISSUE THREAD

**Issue by JackMyers001** [gocd/gocd#10592]
##### Issue Type

- Bug Report

##### Summary

TL;DR agent config files appear to require CRLF line endings (at least under Windows).

###### Full issue details

I have created a Windows Server GoCD Agent container, as one currently does not exist (which is a separate issue...). I need one as I have some Windows-only software that I want to run through GoCD and the Kubernetes plugin for GoCD.

When the Kubernetes plugin creates a Pod, it sets a bunch of environment variables in the container (`GO_EA_SERVER_URL`, `GO_EA_AUTO_REGISTER_KEY`, `GO_EA_AUTO_REGISTER_ELASTIC_PLUGIN_ID`, `GO_EA_AUTO_REGISTER_ELASTIC_AGENT_ID` etc.)

When my Windows container runs, it launches a PowerShell script that writes these environment variables to their respective files (`auto-register.properties` and `wrapper-properties.conf`).

However, I have discovered that GoCD agents (at least under Windows) ***require*** that these files are written with CRLF line endings. If they aren't, only the first line will be read in...?

For example, in my PowerShell script I *tried* to have:

```powershell
@"
agent.auto.register.key=$env:GO_EA_AUTO_REGISTER_KEY
agent.auto.register.elasticAgent.agentId=$env:GO_EA_AUTO_REGISTER_ELASTIC_AGENT_ID
agent.auto.register.elasticAgent.pluginId=$env:GO_EA_AUTO_REGISTER_ELASTIC_PLUGIN_ID
"@ > $GOCD_CONFIG_DIR\autoregister.properties
```

However, only the first line would be read in, so my container would register as a Static Agent, not an Elastic Agent!

The fix ended up being to change the script to:

```powershell
$autoregister_contents = "`r`nagent.auto.register.key=$env:GO_EA_AUTO_REGISTER_KEY`r`n
agent.auto.register.environments=$env:GO_EA_AUTO_REGISTER_ENVIRONMENT`r`n
agent.auto.register.elasticAgent.pluginId=$env:GO_EA_AUTO_REGISTER_ELASTIC_PLUGIN_ID`r`n
agent.auto.register.elasticAgent.agentId=$env:GO_EA_AUTO_REGISTER_ELASTIC_AGENT_ID";

[System.IO.File]::WriteAllLines("$GOCD_CONFIG_DIR\autoregister.properties", $autoregister_contents)
```

*Note that the `` `r`n `` is CRLF in PowerShell (`` ` `` is the escape char).*

This fix is based on [some lines in the Azure Elastic Agent plugin](https://github.com/gocd/azure-elastic-agent-plugin/blob/fb7de497dee4d46b66fd6b65bc01b7422c2cdd43/src/main/resources/windows_install_go_agent.template.flth#L51-L58). Additionally, this doesn't appear to be an issue on Linux, as seen in the [`docker-entrypoint.sh` for the CentOS 8 GoCD Agent](https://github.com/gocd/docker-gocd-agent-centos-8/blob/260a5b6ac75ecef4801360297fd166b395e9bd86/docker-entrypoint.sh#L27-L33).

If this is intended (which I *highly* doubt), it should be documented somewhere. Otherwise, GoCD agents should be able to handle LF and CRLF config files.

###### Basic environment details

- GoCD Server 22.1.0 (official Docker container)
- Windows 11 & Windows Server 2022 & Windows Server Core 2022 Container agents

**Comment by chadlwilson** [comment]
Are you sure that `wrapper-properties.conf` exhibits the same behaviour with the line endings? The example above seems to only be with respect to the `autoregister.properties`.

These two files are loaded by two different mechanisms (pieces of software, technically) so there's no _particular_ reason to believe that they would have the same line-ending behaviour.

**Comment by JackMyers001** [comment]
I think you are correct with the UTF-16 LE issue; on my Windows 11 machine running PowerShell 5 I see that with the `@"text"@` method I get the same result as you (and with the `[System.IO.File]::WriteAllLines` method I get a UTF-8 file).

Interestingly, with PowerShell 7.2 I get regular UTF-8 text with both methods...

> Are you sure that `wrapper-properties.conf` exhibits the same behaviour with the line endings? The example above seems to only be with respect to the `autoregister.properties`.

I omitted the `wrapper-properties.conf` code as it did basically the exact same thing:

```powershell
@"
wrapper.app.parameter.100=-serverUrl
wrapper.app.parameter.101=$env:GO_EA_SERVER_URL
"@ | Out-File -FilePath "$GOCD_WRAPPER_CONFIG_DIR\wrapper-properties.conf"
```

became

```powershell
$wrapper_properties_contents = "`r`nwrapper.app.parameter.100=-serverUrl`r`n
wrapper.app.parameter.101=$env:GO_EA_SERVER_URL";

[System.IO.File]::WriteAllLines("$GOCD_WRAPPER_CONFIG_DIR\wrapper-properties.conf", $wrapper_properties_contents)
```

Having said that, I *should* have mentioned it because I ***DID*** experience different behavior. In my Dockerfile, I run the `go-agent.bat` directly:

```powershell
Start-Process -FilePath "$GOCD_INSTALL_DIR\bin\go-agent.bat" -Wait -NoNewWindow
```

HOWEVER, until I "figured out" the CRLF (actually UTF-16...) issue, I was running the GoCD Agent like:

```powershell
Start-Process -FilePath "$GOCD_INSTALL_DIR\bin\go-agent.bat" -ArgumentList @("-serverUrl", "$env:GO_EA_SERVER_URL") -Wait -NoNewWindow
```

I had discovered that even though I was writing the `wrapper-properties.conf` file, the GoCD agent would start up and immediately stop because it complained that no parameters were passed (and it requires at least the `serverUrl` parameter). This ***isn't*** as if it was reading the first line of the file and not the second (i.e. it was reading in the `-serverUrl` but complaining that there was no accompanying URL parameter), it was actually not reading in the file ***at all***. The argument list was `[]`.

I still feel as if GoCD agents + the wrapper should be able to handle files in any encoding (they're written in Java, right? I would expect that to handle any encoding with no issue...), so I won't close this issue yet. At the very least, the docs should be updated to mention that the agent config files must be UTF-8 encoded.

**Comment by chadlwilson** [comment]
The problem of "what encoding is this file?" is unsolvable in the general sense - some tools (editors generally) tend to use heuristics to try and pick the "best" or "most likely" encoding based on the first 'n' chars (or the entire file content in some cases), where even that becomes difficult if you are talking about UTF-8 vs ASCII vs CP-1252 vs ISO-8859-1, or UTF-16 encodings without BOMs.

By default Java tends to pick platform default or assumes UTF-8 depending on which mechanism you are using to read text, or you tell it when reading a char file what the encoding/charset should be (UTF-8 in this case).

The wrappers are third party code, mainly C, implemented differently for different platforms. They seem to take the approach of allowing a comment to [specify the encoding at the top of the file](https://wrapper.tanukisoftware.com/doc/english/props-directives.html).  Otherwise I think they also assume UTF-8, but the the file is in fact UTF-16 LE w/ BOM how they degrade/interpret it probably becomes indeterminate, likely it throws most away as invalid due to all the `0x00`s from UTF-16.

Nevertheless, finding the appropriate place to document is probably worthwhile. If you have suggestions of where you were looking (or would have looked), that'd be useful.

**Comment by JackMyers001** [comment]
> some tools (editors generally) tend to use heuristics to try and pick the "best" or "most likely" encoding based on the first 'n' chars (or the entire file content in some cases)

This is what I would have expected Java to do, to be honest. Still, if the requirement for UTF-8 is documented then it's a non-issue.

> The wrappers are third party code, mainly C, implemented differently for different platforms

I can see why the wrapper wasn't loading in any parameters then! String manipulation in C, fun times...

> If you have suggestions of where you were looking (or would have looked), that'd be useful.

The most obvious place to put the info for `autoregister.properties` would be on the [doc page titled "Auto registration of remote agents](https://docs.gocd.org/current/advanced_usage/agent_auto_register.html); the second bullet point talks about creating the file.

As for the `wrapper-properties.conf`, I would put a note in the ["Installing GoCD Agent on Linux"](https://docs.gocd.org/current/installation/install/agent/linux.html), under the "Configuring the GoCD agent" section.

There is an [equivalent Windows page](https://docs.gocd.org/current/installation/install/agent/windows.html), however there is no section detailing editing of config files. A small section on editing config files would be good.

Also there is a [page for installing GoCD agents using the zip installer](https://docs.gocd.org/current/installation/install/agent/zip.html) that also mentions editing the `wrapper-properties.conf`, so there too?

Additionally, it might be worth adding a comment to the auto-generated `wrapper-properties.conf` that mentions that the file must be UTF-8. The file already looks like (from a fresh install of the GoCD agent):

```text
; This file was automatically generated by the gocd agent installer
; See wrapper-properties.conf.example for passing options and environment variables to the gocd agent
wrapper.app.parameter.100=-serverUrl
wrapper.app.parameter.101=http://MY-GOCD-IP:8153/go
```

I don't think it's unreasonable to add an extra comment saying "If you edit this file, make sure it is UTF-8 encoded" or something similar.

Finally, a not-so-obvious spot might be to add a comment to the [previously-mentioned Azure Elastic Agent plugin](https://github.com/gocd/azure-elastic-agent-plugin/blob/fb7de497dee4d46b66fd6b65bc01b7422c2cdd43/src/main/resources/windows_install_go_agent.template.flth#L51-L58), as it's not obvious why it writes the file the way it does.

On a slightly unrelated note, if I can get my employer to give the green light, I'd like to release my Windows Server Container Dockerfile + PowerShell entrypoint script as an official GoCD agent container (or at least as a template that others can reference). Where would be the best place to open an issue for that conversation to happen? This repo? The Kubernetes plugin repo? All the other Linux-based containers (e.g. CentOS, Ubuntu) have their own separate repo.

**Comment by chadlwilson** [comment]
> This is what I would have expected Java to do, to be honest. Still, if the requirement for UTF-8 is documented then it's a non-issue.

I'm not sure I have seen any language do this out of the box. It's largely indeterminate, and SDK's don't tend to like doing indeterminate "magic" things. If you have an example/reference from another language's standard SDK that you can think of, I'd be interested in looking at their approach.

On the docs, if you're interested in helping, PRs are welcome at https://github.com/gocd/docs.go.cd - it's largely just markdown. (See the "suggest edits to this page" link at the bottom for quick links)

> On a slightly unrelated note, if I can get my employer to give the green light, I'd like to release my Windows Server Container Dockerfile + PowerShell entrypoint script as an official GoCD agent container (or at least as a template that others can reference). Where would be the best place to open an issue for that conversation to happen? This repo? The Kubernetes plugin repo? All the other Linux-based containers (e.g. CentOS, Ubuntu) have their own separate repo.

Those repos are maintained automatically from the code within this repo, based on templates (in `buildsrc/main/resources` and build automation within the `docker/` subfolder. With the resources the project has available right now being very limited, I doubt there's much appetite to officially support additional images right now, especially Windows containers which come with all their own taxes as you can't run them/test them anywhere but... Windows :-). The current smattering of Linux-based images is difficult enough :-(

The closest we already have (which I would point you to if I was aware you were trying to do this or if I'd seen requests) would be at https://github.com/gocd-contrib/gocd-oss-cookbooks - this [Dockerfile](https://github.com/gocd-contrib/gocd-oss-cookbooks/blob/master/Dockerfile.windowsservercore2019) in particular, however this uses some golang-based bootstrapper that I don't have context for rather than the standard Windows bootstrapper. The images here are used to create agents for building/testing GoCD itself on Windows.

**Comment by chadlwilson** [comment]
Rather than updating all those individual docs pages for the wrapper properties (there are dozens of references) I've added the explicit encoding directive to all the sample files included with GoCD itself. Other than that, the agent docs are updated at https://docs.gocd.org/current/advanced_usage/agent_auto_register.html

Feel free to submit PRs for other docs locations relevant to wrappers if you feel it is really necessary.
